key.go

```diff
@@ -0,0 +1,810 @@
+// The values are taken from https://www.rfc-editor.org/rfc/rfc7517#section-4.3
+func KeyOpFromString(val string) (KeyOp, error) {
+	switch val {
+	case "sign":
+		return KeyOpSign, nil
+	case "verify":
+		return KeyOpVerify, nil
+	case "encrypt":
+		return KeyOpEncrypt, nil
+	case "decrypt":
+		return KeyOpDecrypt, nil
+	case "wrapKey":
+		return KeyOpWrapKey, nil
+	case "unwrapKey":
+		return KeyOpUnwrapKey, nil
+	case "deriveKey":
+		return KeyOpDeriveKey, nil
+	case "deriveBits":
+		return KeyOpDeriveBits, nil
+	default:
+		return KeyOpInvalid, fmt.Errorf("unknown key_ops value %q", val)
+	}
+}
+
+func (ko KeyOp) String() string {
+	switch ko {
+	case KeyOpSign:
+		return "sign"
+	case KeyOpVerify:
+		return "verify"
+	case KeyOpEncrypt:
+		return "encrypt"
+	case KeyOpDecrypt:
+		return "decrypt"
+	case KeyOpWrapKey:
+		return "wrapKey"
+	case KeyOpUnwrapKey:
+		return "unwrapKey"
+	case KeyOpDeriveKey:
+		return "deriveKey"
+	case KeyOpDeriveBits:
+		return "deriveBits"
+	case KeyOpMACCreate:
+		return "MAC create"
+	case KeyOpMACVerify:
+		return "MAC verify"
+	default:
+		return "unknown key_op value " + strconv.Itoa(int(ko))
+	}
+}
+
+func (ko KeyOp) IsSupported() bool {
+	return ko >= 1 && ko <= 10
+}
+
+// MarshalCBOR marshals the KeyOp as a CBOR int.
+func (ko KeyOp) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int64(ko))
+}
+
+// UnmarshalCBOR populates the KeyOp from the provided CBOR value (must be int
+// or tstr).
+func (ko *KeyOp) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key_ops value %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := KeyOpFromString(raw.String())
+		if err != nil {
+			return err
+		}
+
+		*ko = v
+	} else {
+		v := raw.Int()
+		*ko = KeyOp(v)
+
+		if !ko.IsSupported() {
+			return fmt.Errorf("unknown key_ops value %d", v)
+		}
+	}
+
+	return nil
+}
+
+// KeyType identifies the family of keys represented by the associated Key.
+// This determines which files within the Key must be set in order for it to be
+// valid.
+type KeyType int64
+
+const (
+	// Invlaid key type
+	KeyTypeInvalid KeyType = 0
+	// Octet Key Pair
+	KeyTypeOkp KeyType = 1
+	// Elliptic Curve Keys w/ x- and y-coordinate pair
+	KeyTypeEc2 KeyType = 2
+	// Symmetric Keys
+	KeyTypeSymmetric KeyType = 4
+)
+
+// KeyTypeFromString returns the KeyType corresponding to the specified name.
+func KeyTypeFromString(v string) (KeyType, error) {
+	switch v {
+	case "OKP":
+		return KeyTypeOkp, nil
+	case "EC2":
+		return KeyTypeEc2, nil
+	case "Symmetric":
+		return KeyTypeSymmetric, nil
+	default:
+		return KeyTypeInvalid, fmt.Errorf("unknown key type value %q", v)
+	}
+}
+
+func (kt KeyType) String() string {
+	switch kt {
+	case KeyTypeOkp:
+		return "OKP"
+	case KeyTypeEc2:
+		return "EC2"
+	case KeyTypeSymmetric:
+		return "Symmetric"
+	default:
+		return "unknown key type value " + strconv.Itoa(int(kt))
+	}
+}
```

Comment:
Similar to `Algorithm`, the `Name` field of the `KeyType` should not be used as `Value`.

It is stated in [RFC 8152 Section 16.7](https://datatracker.ietf.org/doc/html/rfc8152#section-16.7):
>    Name:  This is a descriptive name that enables easier reference to
>       the item.  The name MUST be unique.  **It is not used in the
>       encoding.**
> 
>    Value:  This is the value used to identify the curve.  These values
>       MUST be unique.  The value can be a positive integer, a negative
>       integer, or a string.